From 46c978f4e8c4fc12d3f5d640d3cf156dd98a5463 Mon Sep 17 00:00:00 2001
From: Patrice Chotard <patrice.chotard@foss.st.com>
Date: Fri, 13 Jun 2025 08:54:52 +0200
Subject: [PATCH] memory: stm32-omi: Optimize FIFO accesses using u16 or u32

FIFO accesses uses u8 only for read/write.
In order to optimize throughput, add u16 or u32 read/write
accesses when possible.

write 64MB	before 	after	ratio
sNOR   84MHz : 	78 sec	68 sec  -13%
sNAND 133MHz : 	10sec	10 sec	0%

Change-Id: I0e9ecc30511cf40422aae3fcdd31432f9026f611
Signed-off-by: Patrice Chotard <patrice.chotard@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/469046
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/memory/stm32-omi.c   | 45 +++++++++++++++++++++++++++++-------
 drivers/spi/spi-stm32-ospi.c |  4 ++--
 include/memory/stm32-omi.h   |  2 +-
 3 files changed, 40 insertions(+), 11 deletions(-)

--- a/drivers/memory/stm32-omi.c
+++ b/drivers/memory/stm32-omi.c
@@ -6,14 +6,32 @@
 
 #include <memory/stm32-omi.h>
 
-static void stm32_omi_read_fifo(u8 *val, void __iomem *addr)
+static void stm32_omi_read_fifo(void *val, void __iomem *addr, u8 len)
 {
-	*val = readb_relaxed(addr);
+	switch (len) {
+	case sizeof(u32):
+		*((u32 *)val) = readl_relaxed(addr);
+		break;
+	case sizeof(u16):
+		*((u16 *)val) = readw_relaxed(addr);
+		break;
+	case sizeof(u8):
+		*((u8 *)val) = readb_relaxed(addr);
+	};
 }
 
-static void stm32_omi_write_fifo(u8 *val, void __iomem *addr)
+static void stm32_omi_write_fifo(void *val, void __iomem *addr, u8 len)
 {
-	writeb_relaxed(*val, addr);
+	switch (len) {
+	case sizeof(u32):
+		writel_relaxed(*((u32 *)val), addr);
+		break;
+	case sizeof(u16):
+		writew_relaxed(*((u16 *)val), addr);
+		break;
+	case sizeof(u8):
+		writeb_relaxed(*((u8 *)val), addr);
+	};
 }
 
 int stm32_omi_abort(struct stm32_omi *omi)
@@ -37,19 +55,20 @@ int stm32_omi_abort(struct stm32_omi *om
 }
 EXPORT_SYMBOL(stm32_omi_abort);
 
-int stm32_omi_tx_poll(struct stm32_omi *omi, u8 *buf, u32 len, bool read)
+int stm32_omi_tx_poll(struct stm32_omi *omi, void *buf, u32 len, bool read)
 {
 	void __iomem *regs_base = omi->regs_base;
-	void (*tx_fifo)(u8 *val, void __iomem *addr);
+	void (*tx_fifo)(void *val, void __iomem *addr, u8 len);
 	u32 sr;
 	int ret;
+	u8 step;
 
 	if (read)
 		tx_fifo = stm32_omi_read_fifo;
 	else
 		tx_fifo = stm32_omi_write_fifo;
 
-	while (len--) {
+	while (len) {
 		ret = readl_relaxed_poll_timeout_atomic(regs_base + OSPI_SR,
 							sr, sr & SR_FTF, 1,
 							STM32_FIFO_TIMEOUT_US);
@@ -60,7 +79,17 @@ int stm32_omi_tx_poll(struct stm32_omi *
 					len, sr);
 			return ret;
 		}
-		tx_fifo(buf++, regs_base + OSPI_DR);
+
+		if (len >= sizeof(u32))
+			step = sizeof(u32);
+		else if (len >= sizeof(u16))
+			step = sizeof(u16);
+		else
+			step = sizeof(u8);
+
+		tx_fifo(buf, regs_base + OSPI_DR, step);
+		len -= step;
+		buf += step;
 	}
 
 	return 0;
--- a/drivers/spi/spi-stm32-ospi.c
+++ b/drivers/spi/spi-stm32-ospi.c
@@ -144,7 +144,7 @@ static int stm32_ospi_tx(struct stm32_os
 {
 	struct stm32_omi *omi = ospi->omi;
 	struct stm32_ospi_flash *flash = &ospi->flash[ospi->last_cs];
-	u8 *buf;
+	void *buf;
 	u8 dummy = 0xff;
 	int ret;
 
@@ -162,7 +162,7 @@ static int stm32_ospi_tx(struct stm32_os
 	if (op->data.dir == SPI_MEM_DATA_IN)
 		buf = op->data.buf.in;
 	else
-		buf = (u8 *)op->data.buf.out;
+		buf = (void *)op->data.buf.out;
 
 	if (flash->octal_dtr && op->addr.val % 2) {
 		/* Read/write dummy byte */
--- a/include/memory/stm32-omi.h
+++ b/include/memory/stm32-omi.h
@@ -195,7 +195,7 @@ void stm32_omi_dlyb_stop(struct stm32_om
 void stm32_omi_dma_callback(void *arg);
 int stm32_omi_dma_setup(struct stm32_omi *omi,
 			struct dma_slave_config *dma_cfg);
-int stm32_omi_tx_poll(struct stm32_omi *omi, u8 *buf, u32 len, bool read);
+int stm32_omi_tx_poll(struct stm32_omi *omi, void *buf, u32 len, bool read);
 int stm32_omi_wait_cmd(struct stm32_omi *omi);
 int stm32_omi_wait_nobusy(struct stm32_omi *omi);
 
